File: src/mcp/tools/relationship.ts
// src/mcp/tools/relationship.ts
// MCP tools: create_relationship, update_relationship, delete_relationship

import { z } from 'zod'
import { assertProjectAccess, getAuthedUserId } from '../auth'
import { socketEmitWithAck } from '../socket-manager'
import {
  McpError,
  ackCodeToMcpCode,
  makeMcpError,
  makeMcpSuccess,
  toMcpErrorResponse,
} from '../errors'
import type { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js'
import { cardinalitySchema } from '@/data/schema'
import {
  getRelationshipProjectId,
  getWhiteboardProjectId,
} from '@/data/resolve-project'
import { findRelationshipById } from '@/data/relationship'


export function registerRelationshipTools(server: McpServer): void {
  // -------------------------------------------------------------------------
  // create_relationship
  // -------------------------------------------------------------------------
  server.tool(
    'create_relationship',
    'Create a relationship between two tables.',
    {
      whiteboardId: z.string().uuid().describe('The whiteboard UUID'),
      sourceTableId: z.string().uuid().describe('Source table UUID'),
      targetTableId: z.string().uuid().describe('Target table UUID'),
      sourceColumnId: z.string().uuid().describe('Source column UUID'),
      targetColumnId: z.string().uuid().describe('Target column UUID'),
      cardinality: cardinalitySchema.describe('Relationship cardinality'),
      label: z.string().max(255).optional().describe('Optional label'),
    },
    async ({
      whiteboardId,
      sourceTableId,
      targetTableId,
      sourceColumnId,
      targetColumnId,
      cardinality,
      label,
    }) => {
      try {
        const userId = getAuthedUserId()
        const projectId = await getWhiteboardProjectId(whiteboardId)
        if (!projectId)
          throw new McpError(
            'NOT_FOUND',
            `Whiteboard ${whiteboardId} not found.`,
          )
        await assertProjectAccess(userId, projectId)

        const payload: Record<string, unknown> = {
          sourceTableId,
          targetTableId,
          sourceColumnId,
          targetColumnId,
          cardinality,
        }
        if (label !== undefined) payload.label = label

        const ack = await socketEmitWithAck(
          whiteboardId,
          'relationship:create',
          payload,
        )
        if (ack.ok === false) {
          return makeMcpError(
            ackCodeToMcpCode(ack.code as string),
            (ack.message as string | undefined) ?? 'Server rejected relationship creation.',
          )
        }
        return makeMcpSuccess(ack.entity)
      } catch (err) {
        return toMcpErrorResponse(err)
      }
    },
  )

  // -------------------------------------------------------------------------
  // update_relationship
  // -------------------------------------------------------------------------
  server.tool(
    'update_relationship',
    "Update a relationship's cardinality, label, or endpoints. " +
      'The server validates referential integrity on all endpoint changes.',
    {
      relationshipId: z.string().uuid().describe('The relationship UUID'),
      sourceTableId: z
        .string()
        .uuid()
        .optional()
        .describe('New source table UUID'),
      targetTableId: z
        .string()
        .uuid()
        .optional()
        .describe('New target table UUID'),
      sourceColumnId: z
        .string()
        .uuid()
        .optional()
        .describe('New source column UUID'),
      targetColumnId: z
        .string()
        .uuid()
        .optional()
        .describe('New target column UUID'),
      cardinality: cardinalitySchema.optional().describe('New cardinality'),
      label: z.string().max(255).optional().describe('New label'),
    },
    async ({
      relationshipId,
      sourceTableId,
      targetTableId,
      sourceColumnId,
      targetColumnId,
      cardinality,
      label,
    }) => {
      try {
        const userId = getAuthedUserId()
        const projectId = await getRelationshipProjectId(relationshipId)
        if (!projectId)
          throw new McpError(
            'NOT_FOUND',
            `Relationship ${relationshipId} not found.`,
          )
        await assertProjectAccess(userId, projectId)

        const existing = await findRelationshipById(relationshipId)
        if (!existing)
          throw new McpError(
            'NOT_FOUND',
            `Relationship ${relationshipId} not found.`,
          )

        const payload: Record<string, unknown> = { relationshipId }
        if (sourceTableId !== undefined) payload.sourceTableId = sourceTableId
        if (targetTableId !== undefined) payload.targetTableId = targetTableId
        if (sourceColumnId !== undefined)
          payload.sourceColumnId = sourceColumnId
        if (targetColumnId !== undefined)
          payload.targetColumnId = targetColumnId
        if (cardinality !== undefined) payload.cardinality = cardinality
        if (label !== undefined) payload.label = label

        const ack = await socketEmitWithAck(
          existing.whiteboardId,
          'relationship:update',
          payload,
        )
        if (ack.ok === false) {
          return makeMcpError(
            ackCodeToMcpCode(ack.code as string),
            (ack.message as string | undefined) ?? 'Server rejected relationship update.',
          )
        }
        return makeMcpSuccess(ack.entity)
      } catch (err) {
        return toMcpErrorResponse(err)
      }
    },
  )

  // -------------------------------------------------------------------------
  // delete_relationship
  // -------------------------------------------------------------------------
  server.tool(
    'delete_relationship',
    'Delete a relationship by ID.',
    {
      relationshipId: z.string().uuid().describe('The relationship UUID'),
    },
    async ({ relationshipId }) => {
      try {
        const userId = getAuthedUserId()
        const projectId = await getRelationshipProjectId(relationshipId)
        if (!projectId)
          throw new McpError(
            'NOT_FOUND',
            `Relationship ${relationshipId} not found.`,
          )
        await assertProjectAccess(userId, projectId)

        const existing = await findRelationshipById(relationshipId)
        if (!existing)
          throw new McpError(
            'NOT_FOUND',
            `Relationship ${relationshipId} not found.`,
          )

        const ack = await socketEmitWithAck(
          existing.whiteboardId,
          'relationship:delete',
          { relationshipId },
        )
        if (ack.ok === false) {
          return makeMcpError(
            ackCodeToMcpCode(ack.code as string),
            (ack.message as string | undefined) ?? 'Server rejected relationship deletion.',
          )
        }
        return makeMcpSuccess({ id: relationshipId })
      } catch (err) {
        return toMcpErrorResponse(err)
      }
    },
  )
}

File: internal/tools/relationship.go
package tools

import (
	"context"
	"fmt"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"github.com/LizardLiang/liz-whiteboard-mcp/internal/auth"
	"github.com/LizardLiang/liz-whiteboard-mcp/internal/data"
	mcperr "github.com/LizardLiang/liz-whiteboard-mcp/internal/errors"
	"github.com/LizardLiang/liz-whiteboard-mcp/internal/schema"
	"github.com/LizardLiang/liz-whiteboard-mcp/internal/socket"
)

type createRelationshipInput struct {
	WhiteboardID   string  `json:"whiteboardId" jsonschema:"The whiteboard UUID"`
	SourceTableID  string  `json:"sourceTableId" jsonschema:"Source table UUID"`
	TargetTableID  string  `json:"targetTableId" jsonschema:"Target table UUID"`
	SourceColumnID string  `json:"sourceColumnId" jsonschema:"Source column UUID"`
	TargetColumnID string  `json:"targetColumnId" jsonschema:"Target column UUID"`
	Cardinality    string  `json:"cardinality" jsonschema:"Relationship cardinality"`
	Label          *string `json:"label,omitempty" jsonschema:"Optional label"`
}

type updateRelationshipInput struct {
	RelationshipID string  `json:"relationshipId" jsonschema:"The relationship UUID"`
	SourceTableID  *string `json:"sourceTableId,omitempty" jsonschema:"New source table UUID"`
	TargetTableID  *string `json:"targetTableId,omitempty" jsonschema:"New target table UUID"`
	SourceColumnID *string `json:"sourceColumnId,omitempty" jsonschema:"New source column UUID"`
	TargetColumnID *string `json:"targetColumnId,omitempty" jsonschema:"New target column UUID"`
	Cardinality    *string `json:"cardinality,omitempty" jsonschema:"New cardinality"`
	Label          *string `json:"label,omitempty" jsonschema:"New label"`
}

type relationshipIDInput struct {
	RelationshipID string `json:"relationshipId" jsonschema:"The relationship UUID"`
}

// RegisterRelationshipTools registers create_relationship, update_relationship,
// delete_relationship.
func RegisterRelationshipTools(s *mcp.Server) {
	registerCreateRelationship(s)
	registerUpdateRelationship(s)
	registerDeleteRelationship(s)
}

func registerCreateRelationship(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name:        "create_relationship",
		Description: "Create a relationship between two tables.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in createRelationshipInput) (*mcp.CallToolResult, any, error) {
		for field, val := range map[string]string{
			"whiteboardId":   in.WhiteboardID,
			"sourceTableId":  in.SourceTableID,
			"targetTableId":  in.TargetTableID,
			"sourceColumnId": in.SourceColumnID,
			"targetColumnId": in.TargetColumnID,
		} {
			if e := validateUUID(field, val); e != nil {
				return fail(e)
			}
		}
		if !schema.IsValidCardinality(in.Cardinality) {
			return validationError("Invalid enum value.", "cardinality")
		}
		// W6: label z.string().max(255)
		if in.Label != nil {
			if e := checkMaxLen("label", *in.Label, 255); e != nil {
				return fail(e)
			}
		}

		userID := auth.UserID(ctx)
		projectID, err := data.GetWhiteboardProjectID(ctx, in.WhiteboardID)
		if err != nil {
			return fail(err)
		}
		if projectID == "" {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Whiteboard %s not found.", in.WhiteboardID))
		}
		if err := auth.AssertSchemaEditAccess(ctx, userID, projectID); err != nil {
			return fail(err)
		}

		payload := map[string]any{
			"sourceTableId":  in.SourceTableID,
			"targetTableId":  in.TargetTableID,
			"sourceColumnId": in.SourceColumnID,
			"targetColumnId": in.TargetColumnID,
			"cardinality":    in.Cardinality,
		}
		if in.Label != nil {
			payload["label"] = *in.Label
		}

		ack, err := socket.SocketEmitWithAck(ctx, in.WhiteboardID, userID, "relationship:create", payload)
		if err != nil {
			return fail(err)
		}
		if !ack.OK() {
			return mcpError(mcperr.AckCodeToMcpCode(ack.Code()), msgOr(ack.Message(), "Server rejected relationship creation."))
		}
		return success(ack.Entity())
	})
}

func registerUpdateRelationship(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name: "update_relationship",
		Description: "Update a relationship's cardinality, label, or endpoints. " +
			"The server validates referential integrity on all endpoint changes.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in updateRelationshipInput) (*mcp.CallToolResult, any, error) {
		if e := validateUUID("relationshipId", in.RelationshipID); e != nil {
			return fail(e)
		}
		for field, val := range map[string]*string{
			"sourceTableId":  in.SourceTableID,
			"targetTableId":  in.TargetTableID,
			"sourceColumnId": in.SourceColumnID,
			"targetColumnId": in.TargetColumnID,
		} {
			if val != nil {
				if e := validateUUID(field, *val); e != nil {
					return fail(e)
				}
			}
		}
		if in.Cardinality != nil && !schema.IsValidCardinality(*in.Cardinality) {
			return validationError("Invalid enum value.", "cardinality")
		}
		// W6: label z.string().max(255)
		if in.Label != nil {
			if e := checkMaxLen("label", *in.Label, 255); e != nil {
				return fail(e)
			}
		}

		userID := auth.UserID(ctx)
		projectID, err := data.GetRelationshipProjectID(ctx, in.RelationshipID)
		if err != nil {
			return fail(err)
		}
		if projectID == "" {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Relationship %s not found.", in.RelationshipID))
		}
		if err := auth.AssertSchemaEditAccess(ctx, userID, projectID); err != nil {
			return fail(err)
		}

		existing, err := data.FindRelationshipByID(ctx, in.RelationshipID)
		if err != nil {
			return fail(err)
		}
		if existing == nil {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Relationship %s not found.", in.RelationshipID))
		}

		payload := map[string]any{"relationshipId": in.RelationshipID}
		if in.SourceTableID != nil {
			payload["sourceTableId"] = *in.SourceTableID
		}
		if in.TargetTableID != nil {
			payload["targetTableId"] = *in.TargetTableID
		}
		if in.SourceColumnID != nil {
			payload["sourceColumnId"] = *in.SourceColumnID
		}
		if in.TargetColumnID != nil {
			payload["targetColumnId"] = *in.TargetColumnID
		}
		if in.Cardinality != nil {
			payload["cardinality"] = *in.Cardinality
		}
		if in.Label != nil {
			payload["label"] = *in.Label
		}

		ack, err := socket.SocketEmitWithAck(ctx, existing.WhiteboardID, userID, "relationship:update", payload)
		if err != nil {
			return fail(err)
		}
		if !ack.OK() {
			return mcpError(mcperr.AckCodeToMcpCode(ack.Code()), msgOr(ack.Message(), "Server rejected relationship update."))
		}
		return success(ack.Entity())
	})
}

func registerDeleteRelationship(s *mcp.Server) {
	mcp.AddTool(s, &mcp.Tool{
		Name:        "delete_relationship",
		Description: "Delete a relationship by ID.",
	}, func(ctx context.Context, _ *mcp.CallToolRequest, in relationshipIDInput) (*mcp.CallToolResult, any, error) {
		if e := validateUUID("relationshipId", in.RelationshipID); e != nil {
			return fail(e)
		}
		userID := auth.UserID(ctx)
		projectID, err := data.GetRelationshipProjectID(ctx, in.RelationshipID)
		if err != nil {
			return fail(err)
		}
		if projectID == "" {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Relationship %s not found.", in.RelationshipID))
		}
		if err := auth.AssertSchemaEditAccess(ctx, userID, projectID); err != nil {
			return fail(err)
		}

		existing, err := data.FindRelationshipByID(ctx, in.RelationshipID)
		if err != nil {
			return fail(err)
		}
		if existing == nil {
			return mcpError(mcperr.NotFound, fmt.Sprintf("Relationship %s not found.", in.RelationshipID))
		}

		ack, err := socket.SocketEmitWithAck(ctx, existing.WhiteboardID, userID, "relationship:delete", map[string]any{"relationshipId": in.RelationshipID})
		if err != nil {
			return fail(err)
		}
		if !ack.OK() {
			return mcpError(mcperr.AckCodeToMcpCode(ack.Code()), msgOr(ack.Message(), "Server rejected relationship deletion."))
		}
		return success(map[string]any{"id": in.RelationshipID})
	})
}
